server/dataset/cxg_dataset.py

```diff
@@ -50,12 +55,431 @@ def __init__(self, data_locator, app_config=None):
 
         self._validate_and_initialize()
 
+    @staticmethod
+    def _encode_component(component: str) -> str:
+        return quote(component, safe="")
+
+    @staticmethod
+    def _decode_component(component: str) -> str:
+        try:
+            return unquote(component)
+        except Exception:
+            return component
+
+    @staticmethod
+    def _make_json_safe(value: Any) -> Any:
+        if value is None:
+            return None
+        if isinstance(value, (np.generic,)):
+            return value.item()
+        if isinstance(value, (pd.Timestamp,)):
+            return value.isoformat()
+        if isinstance(value, (np.datetime64,)):
+            return pd.Timestamp(value).isoformat()
+        return value
+
+    def _user_root_uri(self, user_id: str) -> str:
+        return path_join(self.url, self._encode_component(user_id))
+
+    def _ensure_group(self, uri: str) -> None:
+        try:
+            obj_type = tiledb.object_type(uri, ctx=self.tiledb_ctx)
+        except TileDBError:
+            obj_type = None
+
+        if obj_type is None:
+            vfs = tiledb.VFS(ctx=self.tiledb_ctx)
+            if not vfs.is_dir(uri):
+                vfs.create_dir(uri)
+            tiledb.group_create(uri, ctx=self.tiledb_ctx)
+        elif obj_type != "group":
+            raise DatasetAccessError(f"Annotation storage path collision at {uri}")
+
+    def _ensure_user_storage(self, user_id: str) -> str:
+        user_root = self._user_root_uri(user_id)
+        self._ensure_group(user_root)
+        return user_root
+
+    def _annotation_array_uri(self, user_root_uri: str, column_name: str) -> str:
+        return path_join(user_root_uri, self._encode_component(column_name))
+
+    def _ensure_annotation_array(self, array_uri: str, length: int) -> None:
+        if length <= 0:
+            return
+
+        try:
+            obj_type = tiledb.object_type(array_uri, ctx=self.tiledb_ctx)
+        except TileDBError:
+            obj_type = None
+
+        if obj_type == "array":
+            try:
+                with tiledb.open(array_uri, mode="r", ctx=self.tiledb_ctx) as arr:
+                    dim = arr.schema.domain.dim(0)
+                    attr = arr.schema.attr(0)
+                    domain = dim.domain
```

Comment:
Helpful context: the first ~400 lines here introduce a TileDB-backed per-user namespace (`<dataset>/<encoded-user>/<column>`). `_ensure_*` builds the group, `_write_annotation_column` writes categorical codes + metadata, and `_load_annotation_column` hydrates them back into pandas categoricals. Everything else (schema augmentation, rename/delete, geneset JSON) layers on top of that storage layout.